[IMP][18.0] partner_import_helper & product_import_helper: xlsx template, wizard, multi-company and tags

arnaudlayec opened the pull request:
Hello,

**UPDATE 2025-12-11**: also add improvements on `product_import_helper` and move shared methods to `base_import_helper`

In short: this PR adds an XLSX template in `partner_import_helper` and `product_import_helper` that is ready to use from a wizard accessible in Settings/Technical menu.
The methods that read the template and load its data are designed to be inheritable.
The XLSX template itself is designed to be self-supportive, with a *Tutorial* sheet (ignored during import) for integrators and possibly customers.

In `base_import_helper`:
* add a a menu "Import Helper" in Settings/Technical, that can be used by other import module
* the template download action
* the XLSX import/read logic methods. It call methods to be defined in `[partner|product]_import_helper`, which use a `vals_list` var corresponding to the Excel sheets
* button and fields to the existing wizard, formerly only used for report of the import
* a "AI Engine" selection field in the wizard
* openpyxl dependency

In `partner_import_helper`:
* XLSX template in `data` file with 3 sheets:
   * **companies**: to add companies, 1 bank account per company (optional), 1 invoice address (optional), 1 delivery address (optional)
   * **contacts**: to add other contacts in companies, using the `type` freely. It can also add non-company related contacts (1st level)
   * **banks**: to add other bank accounts to companies
* add support of multi-company (fake `company_ref` field in `vals`)  and tags (fake `categories` field)

In `product_import_helper` :
* support of product tags
* support of a `supplier_code` magic field, keeping support of `supplier_id`
* the XLSX template has only 1 interpreted sheet `products`. It has a 2nd sheets (+Tutorial) which is not interpreted, the `supplierinfo`, which only is the template to use to import them from the menu *"Purchase > Configuration > Vendor price list"*

arnaudlayec committed:
[IMP][18.0] partner_import_helper: xlsx template, wizard, add support of multi-company and tags

metaminux commented on partner_import_helper/tests/.~lock.res_partner_import_test.xlsx#:
This file should be deleted

arnaudlayec replied:
it has been, thanks

metaminux commented on account_balance_reset/__manifest__.py:
This module (`account_balance_reset`) is not installable in 18.0...

However, I suppose it doesn't hurt 😉

metaminux commented on partner_import_helper/tests/res_partner_import_test.xlsx:
Where do all those test data come from ? It seems some of them are true data...

arnaudlayec replied:
Correct, it's now anonymized

metaminux commented on partner_import_helper/wizards/import_helper.py:
```diff
@@ -22,6 +24,166 @@
+
     # TODO add support for states
     @api.model
     def _prepare_speedy(self, aiengine='chatgpt'):
```
```suggestion
    def _prepare_speedy(self, aiengine='none'):
```
With default as "chatgpt", this module is unusable without an OpenAI API key in the server configuration file...
And tests are failing because there is no such key defined

In fact, _prepare_speedy is inherited from import_helper_base and the same default is applied...

Of course, changing this is disabling calls to ChatGPT, unless we add a selection field in the new wizard to enable it

arnaudlayec replied:
I think this is a good feedback, I followed your advises in my new commit

metaminux commented on partner_import_helper/wizards/import_helper.py:
```diff
@@ -22,6 +24,166 @@
+    def _get_companies_address_types(self):
+        """ Suffixes of col names for company's address types
+            :return: ['contact', 'invoice', 'delivery', 'other']
+        """
+        return [x[0] for x in self.env['res.partner']._fields['type'].selection]
+    def _get_company_vals_default(self):
```
```suggestion
        return [x[0] for x in self.env['res.partner']._fields['type'].selection]
    
    def _get_company_vals_default(self):
```

metaminux commented on partner_import_helper/wizards/import_helper.py:
```diff
@@ -22,6 +24,166 @@
+        return {
+            'is_company': True,
+            'lang': 'fr_FR',
+            'country_name': 'FRA',
+        }
+    def _get_company_address_vals_default(self):
```
```suggestion
        }
    
    def _get_company_address_vals_default(self):
```

metaminux commented on partner_import_helper/wizards/import_helper.py:
```diff
@@ -22,6 +24,166 @@
+        return {
+            'is_company': False,
+            'lang': 'fr_FR',
+            'country_name': 'FRA',
+        }
+    def _load_sheet_companies(self, speedy, headers, vals_list):
```
```suggestion
        }
    
    def _load_sheet_companies(self, speedy, headers, vals_list):
```

metaminux commented on partner_import_helper/__manifest__.py:
```diff
@@ -13,7 +13,10 @@
     'depends': [
         'import_helper_base',
         'phone_validation',  # would be nice to avoid depending on it ?
```
```suggestion
    'depends': [
        'account',
        'import_helper_base',
        'l10n_fr',
        'phone_validation',  # would be nice to avoid depending on it ?
```

- Must depend on "l10n_fr" (at least, maybe "l10n_fr_siret") to use the field "siret" (and "siren", etc.)
- Must depend on "account" to use the "default" payment terms

metaminux commented on product_import_helper/__manifest__.py:
Why changing this file ?

arnaudlayec replied:
This is a glitch :)

metaminux reviewed (changes requested):
Globally, I think the code is running without error

I really think the use of ChatGPT should be an option.
We need to make a choice here...
Mine would be to add a selection field on the wizard to use it or not (and in the future use other engines)

I also think providing the XLSX file as a template should be accessible from the UI.
Hosting the file in the "static" directory and making a button that returns an "ir.actions.act_url" should do that...

arnaudlayec committed:
Update partner_import_helper/wizards/import_helper.py

Co-authored-by: Guillaume Masson <guillaume.masson@meta-it.fr>

arnaudlayec committed:
Update partner_import_helper/wizards/import_helper.py

Co-authored-by: Guillaume Masson <guillaume.masson@meta-it.fr>

arnaudlayec committed:
Update partner_import_helper/wizards/import_helper.py

Co-authored-by: Guillaume Masson <guillaume.masson@meta-it.fr>

arnaudlayec committed:
Remove external dependencies from manifest

Removed external dependencies section from manifest.

arnaudlayec committed:
Update partner_import_helper/__manifest__.py

Co-authored-by: Guillaume Masson <guillaume.masson@meta-it.fr>

arnaudlayec committed:
- put aiengine as an option (selection field)
- anonymize data of test & template

arnaudlayec committed:
Merge pull request #1 from arnaudlayec/18.0-partner_add_xlsx_review

- put aiengine as an option (selection field)

arnaudlayec committed:
add button to download the template from the UI

arnaudlayec committed:
Merge branch '18.0' into 18.0-add-xlsx

arnaudlayec committed:
Merge pull request #2 from arnaudlayec/18.0-add-xlsx

add button to download the template from the UI

arnaudlayec commented:
Hello @metaminux, thank you for your review
I think those were great feedbacks, it's been updated in the new commits

arnaudlayec committed:
[fix] xml error (download action before the download button)

arnaudlayec committed:
[fix] allow shared contacts & banks with empty value in "company_ref" column

arnaudlayec committed:
[IMP]
- move to `base_import_helper` : the XLSX import/read methods, the `file` field and the download buttons
- improve `product_import_helper` : support of product tags and `supplier_code` magic field. Add the wizard.

arnaudlayec committed:
[fix2] allow shared contacts & banks with empty value in "company_ref" column

arnaudlayec committed:
[fix] don't convert INT to STR when reading XLSX

arnaudlayec committed:
[FIX] set "company_id" on contact's addresses (for sheet "company")

arnaudlayec committed:
[FIX] resolve merge conflict with master

arnaudlayec committed:
Merge pull request #3 from akretion/18.0

[IMP] account_import_helper: re-write module

arnaudlayec committed:
[FIX] re-apply change after conflict solving (just the menu-item's parent)

arnaudlayec committed:
[IMP]
- add support of a method to import variant: fixed price, 1 line of the file = 1 variant
- add support of category code
- fix logs not showing for products

arnaudlayec committed:
requirements.txt

arnaudlayec committed:
clean unused keys in __manifest__

arnaudlayec committed:
add support for product tags

arnaudlayec committed:
[FIX] little issue "ValueError: Invalid field 'tags' on model 'product.product'"

arnaudlayec committed:
[IMP] *import_helper: add 'mail_create_nosubscribe' in context for create()

arnaudlayec committed:
[IMP] *import_helper: add 'mail_create_nosubscribe' in context for create()

alexis-via commented on account_import_helper/wizards/account_chart_import_view.xml:
```diff
@@ -55,16 +55,10 @@
 
 <menuitem
         id="account_chart_import_menu"
-        parent="account_chart_tools_parent_menu"
```
You broke the module account_import_helper !!! This module doesn't depend on import_helper_base, so you cannot change parent="" like you did!

alexis-via replied:
I fixed it myself in my last commit

metaminux replied:
Hello @alexis-via,

Would it make sense to add the dependency on `import_helper_base` in `account_import_helper` instead ?

From what I understand, @arnaudlayec was aiming to ensure consistency across the "import_helper" modules.

arnaudlayec replied:
Hello, yes this is it @metaminux 
By the way I'm just discovering the PR was merged. This was not expected, sorry for the glitch @alexis-via
